src/classes/RD2_OpportunityService.cls

```diff
@@ -36,180 +36,373 @@
+                //update future open Opp only
+                if (opp.CloseDate <= currentDate) {
+                    continue;
+                }
+
+                Boolean hasMatchingInstallment = installments != null && (i + 1 <= installments.size());
+                installmentByOppId.put(opp.Id, hasMatchingInstallment ? installments[i] : null);
+
+                Boolean isChanged = syncOppWithSchedule(opp, installmentByOppId.get(opp.Id), rd.npe03__Donor_Name__c);
+                if (isChanged) {
+                    oppsToUpdate.add(opp);
+                }
+            }
+        }
+
+        List<Database.SaveResult> saveResults = Database.update(oppsToUpdate, false);
+        handleErrors(saveResults, oppsToUpdate);
+
+        updatePaymentRecords(installmentByOppId);
+    }
+
+    /***
+    * @description Returns Opp with Close Date = today (if any) and all future open Opps
+    * - If the Close Date = today, then the Opp can be closed or open
+    * - If the Close Date > today, only open Opps are considered
+    *
+    * @param rd Recurring Donation
+    * @return List<Opportunity>
+    */
+    private List<Opportunity> getCurrentOpps(npe03__Recurring_Donation__c rd) {
+        List<Opportunity> opps = new List<Opportunity>();
+
+        if (rd.npe03__Donations__r.isEmpty()) {
+            return opps;
+        }
+        
+        for (Opportunity opp : rd.npe03__Donations__r) {
+            if ((opp.IsClosed == false && opp.CloseDate > currentDate) || opp.CloseDate == currentDate) {
+                opps.add(opp);
+            }
+        }
+
+        return opps;
+    }
+
+    /***
+    * @description Updates Opportunity if it differs from the related Recurring Donation Schedule installment
+    * @param opp Opportunity
+    * @param installment Recurring Donation Schedule installment
+    * @param donorName Recurring Donation donor name
+    * @return Boolean
+    */
+    private Boolean syncOppWithSchedule(Opportunity opp,  RD2_ScheduleService.Installment installment, String donorName) {  
+        //An installment Opportunity should always have a matching RD Schedule installment.
+        //If the installment is null, this is most probably a programatic error.
+        if (installment == null) {
+            opp.addError(System.Label.RD2_ScheduleInstallmentIsRequired);
+            return true;
+        }    
+
+        Boolean isChanged = false;
+
+        if (opp.Amount != installment.installmentAmount) {
+            opp.Amount = installment.installmentAmount;
+            isChanged = true;
+        }
+
+        if (opp.CloseDate != installment.nextDonationDate) {
+            opp.CloseDate = installment.nextDonationDate;
+            opp.name = buildOpportunityName(donorName, opp);
+            isChanged = true;
+        }
 
-        for (npe03__Recurring_Donation__c r : rds) {
-            paymentMethodByRDId.put(r.Id, r.PaymentMethod__c);
+        if (installment.campaignId != null 
```

Comment:
@force2b , if the open future Opp Campaign is null and the RD has a new Campaign, should the Campaign be propagated to the open Opp as well?
@heidmary ^